src/vision/status_quo/barbara_carefully_dismisses_embedded_future.md

```diff
@@ -0,0 +1,210 @@
+vtables to avoid the code duplication, but then the compiler *definitely* cannot
+inline `Future::poll` into the kernel callbacks.
+
+At this point, Barbara realizes she needs to do some benchmarking. Barbara comes
+up with a sample application -- an app that blinks a led and responds to button
+presses -- and implements it twice. One implementation does not use `Future` at
+all, the other does. In addition to two async drivers (a timer driver and a GPIO
+driver), the `Future`-based app uses Barbara's executor as well as a single
+combinator (that manages the state of the app).
+
+Barbara publishes an
+[analysis](https://github.com/tock/design-explorations/tree/master/size_comparison)
+of the relative sizes of the two app implementations, finding a large percentage
+increase in both code size and RAM usage (note: stack usage was not
+investigated). Barbara concludes that **`core::future::Future` is not suitable
+for highly-resource-constrained environments** due to the amount of code and RAM
+required to implement executors and combinators.
+
+Barbara redesigns the library she is building to use a [different
+concept](https://github.com/tock/design-explorations/tree/master/zst_pointer_async)
+for implementing async APIs in Rust that are much lighter weight. She has moved
+on from `Future` and is refining her [async
+traits](https://github.com/tock/libtock-rs/blob/master/core/platform/src/async_traits.rs)
+instead.
+
+## 🤔 Frequently Asked Questions
+
+* **What are the morals of the story?**
+    * `core::future::Future` isn't suitable for every asynchronous API in Rust.
+      `Future` has a lot of capabilities, such as the ability to spawn
+      dynamically-allocated futures, that are unnecessary in embedded system.
+    * We should look at embedded Rust's relationship with `Future` so we don't
+      fragment the embedded Rust ecosystem. Other embedded crates use `Future`,
+      presumably because they are less space constrained than Barbara's
+      codebase.
+* **Why did you choose *Barbara* to tell this story?**
+    * This story is about someone who is an experienced systems programmer and
+      an experienced Rust developer. All the other characters have "new to Rust"
+      or "new to programming" as a key characteristic.
+ **How would this story have played out differently for the other characters?**
+    * [Alan] would have found the `#![no_std]` crate ecosystem lacking async
+      support. He would have moved forward with a `Future`-based implementation,
+      unaware of its impact on code size and RAM usage.
+    * [Grace] would have handled the issue similarly to Barbara, but may not
+      have tried as hard to use `Future`. Barbara has been paying attention to
+      Rust long enough to know how significant the `Future` trait is in the Rust
+      community and ecosystem.
+    * [Niklaus] would really have struggled. If he asked for help, he probably
```

Comment:
This strikes me as an important point. It's fairly "common knowledge" that `Future`s themselves are very lightweight. This story makes it clear that for some resource constrained environments, the machinery required for getting futures to work can sometimes be too heavy weight. I can imagine there being many who would dismiss Niklaus, saying that `Future`s are lightweight and totally appropriate for embedded development without realizing that this might not always be true.

Reply:
In 2016, Aaron Turon described `Future` as [zero cost](https://aturon.github.io/blog/2016/08/11/futures/#zero-cost), but he elaborated on what that meant, and "minimal impact on code size and RAM usage" was not a part of that definition. The phrase "zero cost" has been repeated in descriptions of `Future` ever since, generally without the context that Aaron provided.

Most of Rust's "zero cost" abstractions have a much smaller (or zero) impact on code size and RAM usage after inlining, so it's easy to assume that `Future` is lighter weight than it is.